lib/puppet/provider/bootstrap_rpm/bootstrap_rpm.rb

```diff
@@ -0,0 +1,200 @@
+
+  def exists?
+    copy_sources
+    write_specfile
+    build_rpm
+
+    (!resource[:symlink] || File.exist?(resource[:symlink])) && !rpm_changed?
+  end
+
+  private
+
+  def base_dir
+    @base_dir ||= Dir.mktmpdir
+  end
+
+  def sources_dir
+    mkdir('SOURCES')
+  end
+
+  def spec_dir
+    mkdir('SPECS')
+  end
+
+  def rpm_dir
+    mkdir('RPMS')
+  end
+
+  def srpm_dir
+    mkdir('SRPMS')
+  end
+
+  def copy_sources
+    FileUtils.copy(resource[:script], File.join(sources_dir, File.basename(resource[:script])))
+  end
+
+  def write_specfile
+    File.open(File.join(spec_dir, "#{resource[:name]}.spec"), 'w') do |file|
+      file.write(spec)
+    end
+  end
+
+  def release
+    if (latest = latest_rpm)
+      rel = rpm('-qp', latest, "--queryformat=%{release}")
+      rel.to_i + 1
+    else
+      1
+    end
+  end
+
+  def build_rpm
+    output = rpmbuild(
+      '-ba',
+      File.join(spec_dir, "#{resource[:name]}.spec"),
+      '--define', "_topdir #{base_dir}",
+      '--define', "name #{resource[:name]}",
+      '--define', "release #{release}"
+    )
+  end
+
+  def copy_rpm
+    FileUtils.copy(built_rpm, resource[:dest])
+  end
+
+  def copy_srpm
+    FileUtils.copy(built_srpm, resource[:dest])
+  end
+
+  def rpm_changed?
+    changed = true
+    return changed unless latest_rpm
+
+    Dir.mktmpdir do |dir|
+      srpm = latest_rpm.gsub('noarch', 'src')
+      temp_srpm = File.join(dir, File.basename(srpm))
+      FileUtils.copy(srpm, temp_srpm)
+
+      Dir.chdir(dir) do
+        output = rpm2cpio(temp_srpm)
+        File.open('srpm.cpio', 'w') { |file| file.write(output) }
+        cpio('-idmv', '--file', 'srpm.cpio')
+      end
+
+      old_spec = File.read(Dir.glob("#{dir}/*.spec")[0])
+      old_source = File.read(Dir.glob("#{dir}/#{File.basename(resource[:script])}")[0])
+
+      new_spec = spec
+      new_source = File.read(resource[:script])
+
+      digest_old_spec = Digest::SHA256.hexdigest(old_spec)
+      digest_new_spec = Digest::SHA256.hexdigest(new_spec)
+
+      digest_old_source = Digest::SHA256.hexdigest(old_source)
+      digest_new_source = Digest::SHA256.hexdigest(new_source)
+
+      changed = (digest_old_spec != digest_new_spec) || (digest_old_source != digest_new_source)
+    end
+
+    changed
+  end
+
+  def built_rpm
+    Dir.glob("#{rpm_dir}/noarch/*#{release}.noarch.rpm")[0]
+  end
+
+  def built_srpm
+    Dir.glob("#{srpm_dir}/*")[0]
+  end
+
+  def link_rpm
+    FileUtils.ln_s(latest_rpm, resource[:symlink], force: true) if resource[:symlink]
+  end
+
+  def latest_rpm
+    rpms = Dir.glob("#{resource[:dest]}/*.noarch.rpm")
+    rpms = rpms.reject { |rpm| rpm.end_with?("latest.noarch.rpm") }
+
+    return false if rpms.empty?
+
+    rpms = rpms.collect do |rpm|
+      version = rpm('-qp', rpm, "--queryformat=%{version}")
+      release = rpm('-qp', rpm, "--queryformat=%{release}")
+
+      {'release' => release, 'rpm' => rpm}
+    end
+
+    rpms.max_by { |rpm| rpm['release'].to_i }['rpm']
```

Comment:
Btw, it would be more efficient to only query the version once you need it:
```suggestion
    latest = rpms.max_by { |name| rpm('-qp', name, "--queryformat=%{release}") }
    rpm('-qp', latest, "--queryformat=%{version}")
```

This may also fix some issues with masking. looks like you use `rpm` both as a local variable and as a method.

Code is untested though, so be aware of that :)